src/TeamSpeak3QueryApi/Query/QueryClient.cs

```diff
@@ -31,128 +30,52 @@ public class QueryClient : IDisposable
-        /// <param name="hostName">The host name of the remote server.</param>
-        public QueryClient(string hostName, Protocol type)
-            : this(hostName, DefaultPort, type)
-        { }
-        /// <summary>Creates a new instance of <see cref="TeamSpeak3QueryApi.Net.QueryClient"/> using the provided host TCP port.</summary>
-        /// <param name="hostName">The host name of the remote server.</param>
-        /// <param name="port">The TCP port of the Query API server.</param>
-        public QueryClient(string hostName, int port, Protocol type)
-        {
-            if (string.IsNullOrWhiteSpace(hostName))
-                throw new ArgumentNullException(nameof(hostName));
-            if (!ValidationHelper.ValidateTcpPort(port))
-                throw new ArgumentOutOfRangeException(nameof(port));
-
-            Host = hostName;
-            Port = port;
-            ConnectionType = type;
-            IsConnected = false;
-            Client = new TcpClient();
-        }
-
-        #endregion
+        protected SshClient _sshClient;
+        protected ShellStream _shell;
+        protected string username;
 
-        /// <summary>Connects to the Query API server.</summary>
-        /// <returns>An awaitable <see cref="Task"/>.</returns>
-        public async Task<CancellationTokenSource> ConnectAsync()
+        public void Disconnect()
         {
-            if(ConnectionType != Protocol.Telnet)
-                throw new InvalidOperationException("ConnectAsync Method without parameters can only be used with telnet Query. Please use ConnectSsh method.");
-
-            await Client.ConnectAsync(Host, Port).ConfigureAwait(false);
-            if (!Client.Connected)
-                throw new InvalidOperationException("Could not connect.");
-
-            _ns = Client.GetStream();
-            _reader = new StreamReader(_ns);
-            _writer = new StreamWriter(_ns) { NewLine = "\n" };
-
-            IsConnected = true;
-            OnConnected?.Invoke(this, EventArgs.Empty);
-
-
-            var headline = await _reader.ReadLineAsync().ConfigureAwait(false);
-            if (headline != "TS3")
-            {
-                throw new QueryProtocolException("Telnet Query isn't a valid Teamspeak Query");
-            }
-            await _reader.ReadLineAsync().ConfigureAwait(false); // Ignore welcome message
-            await _reader.ReadLineAsync().ConfigureAwait(false);
+            if (_cts == null)
+                return;
 
-            return ResponseProcessingLoop();
+            OnDisconnected?.Invoke(this, EventArgs.Empty);
+            _cts.Cancel();
         }
 
-        public CancellationTokenSource ConnectSsh(string username, string password)
+        /// <summary>Connects to the Query API server.</summary>
+        /// <returns>An awaitable <see cref="Task"/>.</returns>
+        public virtual Task<CancellationTokenSource> ConnectAsync()
```

Comment:
This should be abstract or non-existent (it will be there in the child classes).

Reply:
Changed to abstract.